Updated box for Chainlink VRF and Chainlink Price Feeds

PatrickAlphaC opened the pull request.

thodges-gh commented on contracts/PriceConsumerV3.sol:
```diff
@@ -0,0 +1,31 @@
+        ) = priceFeed.latestRoundData();
+        return price;
+    }
+}
```
You'll want to update your editor to include the EOF newline.

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,54 @@
+    
+    bytes32 internal keyHash;
+    uint256 internal fee;
+    address public linkTokenAddress;
```
VRFConsumerBase already provides the LINK token address in the internal variable `LINK`. This can be removed and `LINK` used instead.

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,54 @@
+     * Requests randomness from a user-provided seed
+     */
+    function getRandomNumber(uint256 userProvidedSeed) public returns (bytes32 requestId) {
+        require(LINK.balanceOf(address(this)) >= fee, "Not enough LINK - fill contract with faucet");
```
This check is redundant, since the token transfer will fail if the balance of this contract doesn't meet the transfer amount (the fee) anyway.

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,54 @@
+     * Requests the address of the Chainlink Token on this network 
+     */
+    function getChainlinkToken() public view returns (address) {
+        return linkTokenAddress;
```
```suggestion
        return address(LINK);
```

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,54 @@
+    function fulfillRandomness(bytes32 requestId, uint256 randomness) internal override {
+        randomResult = randomness;
+    }
+}
```
Newline

thodges-gh commented on migrations/3_price_consumer_v3.js:
```diff
@@ -0,0 +1,17 @@
+module.exports = async (deployer, network, [defaultAccount]) => {
+    // Local (development) networks need their own deployment of the LINK
+    // token and the Oracle contract
+    if (!network.startsWith('live')) {
```
Looks like we can get rid of this and just have a single code path.

thodges-gh commented on scripts/random-number-consumer-scripts/request-randomness.js:
```diff
@@ -0,0 +1,17 @@
+const RandomNumberConsumer = artifacts.require('RandomNumberConsumer')
+const math = require("mathjs")
```
You can remove this.

thodges-gh commented on scripts/random-number-consumer-scripts/request-randomness.js:
```diff
@@ -0,0 +1,17 @@
+module.exports = async callback => {
+    const USER_PROVIDED_SEED = process.env.RANDOM_SEED || 1
+    const randomNumberConsumer = await RandomNumberConsumer.deployed()
+    console.log('Creating getting a random number from: ', randomNumberConsumer.address)
```
```suggestion
    console.log('Requesting a random number from: ', randomNumberConsumer.address)
```

thodges-gh commented on test/MyContract_test.js:
```diff
@@ -17,6 +17,8 @@ contract('MyContract', accounts => {
   // the type of data, but will not work on a public testnet.
   // For the latest JobIDs, visit our docs here:
   // https://docs.chain.link/docs/testnet-oracles
+  // Alternativly, visti a node listing service and choose your
```
Let's just point to market.link instead of our docs. That page is no longer maintained.

thodges-gh commented on truffle-config.js:
```diff
@@ -21,6 +21,21 @@ module.exports = {
       // Necessary due to https://github.com/trufflesuite/truffle/issues/3008
       skipDryRun: true,
     },
+    // TODO
```
I really don't like leaving commented-out code in source.

thodges-gh commented on README.md:
```diff
@@ -97,3 +106,8 @@ Using network 'live'.
+## TODO
+
+- Add tests for ChainlinkVRF
+- Add tests for Chainlink Price Feeds
```
I'd like to add a few:
- Refactor tests to use `this` instead of defining contracts with `let`
- Use the Chainlink-published mocks for [MockV3Aggregator](https://github.com/smartcontractkit/chainlink/blob/develop/evm-contracts/src/v0.6/tests/MockV3Aggregator.sol) and [VRFCoordinatorMock](https://github.com/smartcontractkit/chainlink/blob/develop/evm-contracts/src/v0.6/tests/VRFCoordinatorMock.sol)
- Upgrade to Solidity 0.6.12 (to show that you don't _have_ to use 0.6.6)

thodges-gh commented on README.md:
```diff
@@ -48,6 +55,8 @@ There are 3 helper scripts provided with this box in the scripts directory:
 - `request-data.js`
 - `read-contract.js`
 
+In addition, for working with Chainlink Price Feeds and ChainlinkVRF there are folders respectivly. 
```
```suggestion
In addition, for working with Chainlink Price Feeds and ChainlinkVRF there are folders respectively. 
```

PatrickAlphaC commented:
Updated with best practices and added todo list. The only section that I don't understand, is using the `LINK` keyword. Since that is an immutable contract, how could I set the LINK token address on a local chain for testing purposes? 

The `MyContract.sol` doesn't have the LINK pieces.

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,52 @@
+    constructor(address _linkTokenAddress) 
+        VRFConsumerBase(
+            0xdD3782915140c8f3b190B5D67eAc6dc5760C46E9, // VRF Coordinator
+            _linkTokenAddress  // LINK Token
```
@PatrickAlphaC This applies only to this `RandomNumberConsumer` contract. You're setting the value for `LINK` here.

thodges-gh commented on contracts/RandomNumberConsumer.sol:
```diff
@@ -0,0 +1,52 @@
+            _linkTokenAddress  // LINK Token
+        ) public
+    {
+        linkTokenAddress = _linkTokenAddress;
```
And setting this variable is redundant.

thodges-gh commented on package.json.orig:
```diff
@@ -0,0 +1,43 @@
+  "description": "A Chainlink example in a Truffle box",
+  "scripts": {
+    "compile": "npx truffle compile",
+    "console:dev": "npx truffle console --network cldev",
+    "console:live": "npx truffle console --network live",
+    "depcheck": "echo '@chainlink/box' && depcheck --ignore-dirs=build/contracts || true",
+    "solhint": "solhint ./contracts/**/*.sol",
+    "lint": "yarn solhint",
+    "migrate:dev": "npx truffle migrate --reset --network cldev",
+    "migrate:live": "npx truffle migrate --network live",
+    "test": "npx truffle test"
+  },
+  "license": "MIT",
+  "dependencies": {
+    "@chainlink/contracts": "^0.0.11",
+    "@openzeppelin/contracts": "^3.1.0",
+<<<<<<< HEAD
+    "@truffle/hdwallet-provider": "^1.0.40",
+    "mathjs": "^8.1.0"
+=======
+    "@truffle/hdwallet-provider": "^1.2.0"
+>>>>>>> 4219f8f21da822083b100dff85eb3b4b5c3c5e3f
+  },
+  "devDependencies": {
+    "@chainlink/belt": "^0.0.3",
+    "@chainlink/test-helpers": "0.0.5",
+    "@openzeppelin/test-helpers": "^0.5.6",
+    "chai": "^4.2.0",
+<<<<<<< HEAD
+    "depcheck": "^0.9.1",
+    "jest": "^26.6.3",
+    "solhint": "^2.1.0",
+    "truffle": "^5.1.5"
+=======
+    "depcheck": "^1.3.1",
+    "solhint": "^3.3.2",
+    "truffle": "^5.1.57"
+>>>>>>> 4219f8f21da822083b100dff85eb3b4b5c3c5e3f
+  }
+}
```
Remove this file.

PatrickAlphaC commented:
Updated with new:

1. Removed the redundant code
2. Removed the not needed file (created due to merge conflict)
